### PR TITLE
[Jobs] Add scheduled jobs api

### PR DESCRIPTION
**Pull request by lhoestq**
I went with this syntax, let me know what you think:

```
hf jobs scheduled run @hourly ubuntu echo hello world
hf jobs scheduled run "0 * * * *" ubuntu echo hello world
hf jobs scheduled ps -a
hf jobs scheduled inspect <id>
hf jobs scheduled delete <id>
hf jobs scheduled suspend <id>
hf jobs scheduled resume <id>
hf jobs scheduled uv run @weekly train.py
```

and

```python
create_scheduled_job(image="ubuntu", command=["echo", "hello world"], schedule="@hourly")
create_scheduled_job(image="ubuntu", command=["echo", "hello world"], schedule="0 * * * *")
list_scheduled_jobs()
inspect_scheduled_job(scheduled_job_id)
delete_scheduled_job(scheduled_job_id)
suspend_scheduled_job(scheduled_job_id)
resume_scheduled_job(scheduled_job_id)
create_scheduled_uv_job("train.py", schedule="@weekly")
```

TODO:
- [x] tests
- [x] docs

### CODE REVIEW

File: src/huggingface_hub/hf_api.py [hunk 1]
```diff
@@ -10453,6 +10453,525 @@ def run_uv_job(
...
+                script_content = f.read()
+
+            commit_hash = self.upload_file(
+                path_or_fileobj=script_content.encode(),
+                path_in_repo=filename,
+                repo_id=repo_id,
+                repo_type="dataset",
+            ).oid
+
+            script_url = f"https://huggingface.co/datasets/{repo_id}/resolve/{commit_hash}/{filename}"
```
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py]:
```suggestion
            script_url = f"{self.endpoint}/datasets/{repo_id}/resolve/{commit_hash}/{filename}"
```

File: src/huggingface_hub/hf_api.py [hunk 2]
```diff
@@ -10453,6 +10453,525 @@ def run_uv_job(
...
+            repo_url = f"https://huggingface.co/datasets/{repo_id}"
```
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py]:
```suggestion
            repo_url = f"{self.endpoint}/datasets/{repo_id}"
```

File: src/huggingface_hub/hf_api.py [hunk 3]
```diff
@@ -10617,3 +11136,10 @@ def _parse_revision_from_pr_url(pr_url: str) -> str:
 inspect_job = api.inspect_job
 cancel_job = api.cancel_job
 run_uv_job = api.run_uv_job
+schedule_job = api.schedule_job
+list_scheduled_jobs = api.list_scheduled_jobs
+inspect_scheduled_job = api.inspect_scheduled_job
+delete_scheduled_job = api.delete_scheduled_job
+suspend_scheduled_job = api.suspend_scheduled_job
+resume_scheduled_job = api.suspend_scheduled_job
```
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py]:
```suggestion
resume_scheduled_job = api.resume_scheduled_job
```

File: src/huggingface_hub/hf_api.py [hunk 4]
```diff
@@ -10453,6 +10453,525 @@ def run_uv_job(
...
+        else:
+            # Local file - upload to HF
+            script_path = Path(script)
+            filename = script_path.name
+            # Parse repo
+            if _repo:
+                repo_id = _repo
+                if "/" not in repo_id:
+                    repo_id = f"{namespace}/{repo_id}"
+                repo_id = _repo
```
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py]: The last `repo_id = _repo` undoes the namespace prefixing
```suggestion
                repo_id = _repo
                if "/" not in repo_id:
                    repo_id = f"{namespace}/{repo_id}"
```

File: src/huggingface_hub/hf_api.py [hunk 5]
```diff
@@ -10453,6 +10453,525 @@ def run_uv_job(
...
+            namespace (`str`, *optional*):
+                The namespace from where it lists the jobs. Defaults to the current user's namespace.
+
+            token `(Union[bool, str, None]`, *optional*):
+                A valid user access token. If not provided, the locally saved token will be used, which is the
+                recommended authentication method. Set to `False` to disable authentication.
+                Refer to: https://huggingface.co/docs/huggingface_hub/quick-start#authentication.
+        """
+        if namespace is None:
+            namespace = whoami(token=token)["name"]
```
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py]:
```suggestion
            namespace = self.whoami(token=token)["name"]
```

File: src/huggingface_hub/hf_api.py [hunk 6]
```diff
@@ -10453,6 +10453,525 @@ def run_uv_job(
...
+        response = get_session().get(
+            f"{self.endpoint}/api/scheduled-jobs/{namespace}",
+            headers=self._build_hf_headers(token=token),
+            timeout=timeout,
+        )
+        response.raise_for_status()
```
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py]:
```suggestion
        hf_raise_for_status(response)
```

File: src/huggingface_hub/_jobs_api.py [hunk 7]
```diff
@@ -142,3 +142,119 @@ def __init__(self, **kwargs) -> None:
...
+
+    def __init__(self, **kwargs) -> None:
+        self.id = kwargs["id"]
+        self.at = parse_datetime(kwargs["at"])
+
+
+@dataclass
+class ScheduledJobStatus:
+    last_job: Optional[LastJobInfo]
+    next_job_run_at: datetime
```
Comment (hanouticelina) [inline src/huggingface_hub/_jobs_api.py]: If we are not sure this value will always be present, it should be optional and we should add a guard line 192
```suggestion
    next_job_run_at: Optional[datetime]
```

File: src/huggingface_hub/_jobs_api.py [hunk 8]
```diff
@@ -142,3 +142,119 @@ def __init__(self, **kwargs) -> None:
...
+
+    def __init__(self, **kwargs) -> None:
+        last_job = kwargs.get("lastJob") or kwargs.get("last_job")
+        self.last_job = LastJobInfo(**last_job) if last_job else None
+        next_job_run_at = kwargs.get("nextJobRunAt") or kwargs.get("next_job_run_at")
+        self.next_job_run_at = parse_datetime(str(next_job_run_at))
```
Comment (hanouticelina) [inline src/huggingface_hub/_jobs_api.py]:
```suggestion
        self.next_job_run_at = parse_datetime(str(next_job_run_at)) if next_job_run_at else None
```

File: src/huggingface_hub/hf_api.py [hunk 9]
```diff
@@ -10453,6 +10453,525 @@ def run_uv_job(
...
+            >>> inspect_scheduled_job(scheduled_job.id)
+            ```
+        """
+        if namespace is None:
+            namespace = self.whoami(token=token)["name"]
+        response = get_session().get(
+            f"{self.endpoint}/api/scheduled-jobs/{namespace}/{scheduled_job_id}",
+            headers=self._build_hf_headers(token=token),
+        )
+        response.raise_for_status()
```
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py]:
```suggestion
        hf_raise_for_status(response)
```

File: docs/source/en/guides/cli.md [hunk 10]
```diff
@@ -760,3 +760,45 @@ Run UV scripts (Python scripts with inline dependencies) on HF infrastructure:
...
+
+```bash
+# Schedule a job that runs every hour
+>>> hf jobs scheduled run @hourly python:3.12 python -c 'print("This runs every hour!")'
+
+# Use the CRON syntax
+>>> hf jobs scheduled run "*/5 * * * *" python:3.12 python -c 'print("This runs every 5 minutes!")'
+
+# Schedule with GPU
+>>> hf jobs schefuled run @hourly --flavor a10g-small pytorch/pytorch:2.6.0-cuda12.4-cudnn9-devel \
```
Comment (hanouticelina) [inline docs/source/en/guides/cli.md]:
```suggestion
>>> hf jobs scheduled run @hourly --flavor a10g-small pytorch/pytorch:2.6.0-cuda12.4-cudnn9-devel \
```

File: src/huggingface_hub/_jobs_api.py [hunk 11]
```diff
@@ -142,3 +142,119 @@ def __init__(self, **kwargs) -> None:
...
+            Scheduled Job ID.
+        created_at (`datetime` or `None`):
+            When the scheduled Job was created.
+        tags (`List[str]` or `None`):
+            The tags of the scheduled Job.
+        schedule (`str` or `None`):
+            One of "@annually", "@yearly", "@monthly", "@weekly", "@daily", "@hourly", or a
+            CRON schedule expression (e.g., '0 9 * * 1' for 9 AM every Monday).
+        suspend (`bool` or `None`):
+            Whether the the scheduled job is suspended (paused).
```
Comment (hanouticelina) [inline src/huggingface_hub/_jobs_api.py]:
```suggestion
            Whether the scheduled job is suspended (paused).
```

File: src/huggingface_hub/hf_api.py [hunk 12]
```diff
@@ -10453,6 +10453,525 @@ def run_uv_job(
             token=token,
         )
 
+    def schedule_job(
```
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py]: personal preference here, "create" is explicit and pairs naturally with list, inspect and delete
```suggestion
    def create_scheduled_job(
```

**Review by hanouticelina — COMMENTED** [review]
Thanks! I did a first round of review with a couple of comments, mostly nits.
Quick UX question: what's the advantage of adding a dedicated `scheduled` subcommand vs allowing scheduling directly on `hf jobs run` with a `--schedule` flag, e.g:
```
hf jobs run --schedule @hourly python:3.12 python -c 'print("This runs every hour!")'
```
not a strong opinion on that but i think this allows to have one consistent way to use and think about `hf jobs`: call `hf jobs` run for both immediate and recurring runs, and just toggle `--schedule` when you want it scheduled.

**Comment by Wauplin** [comment]
(x-linking moon PR https://github.com/huggingface-internal/moon-landing/pull/14655)

File: src/huggingface_hub/hf_api.py [hunk 13]
```diff
@@ -10453,6 +10453,525 @@ def run_uv_job(
...
+        self,
+        *,
+        image: str,
+        command: List[str],
+        schedule: str,
+        suspend: bool = False,
+        concurrency: bool = False,
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py]:
```suggestion
        suspend: Optional[bool] = None,
        concurrency: Optional[bool] = None,
```

Better to default to `None` and pass a value in the body only if non-none value is passed. This way we let the server handle the default value

File: src/huggingface_hub/hf_api.py [hunk 14]
```diff
@@ -10453,6 +10453,525 @@ def run_uv_job(
...
+
+            command (`List[str]`):
+                The command to run. Example: `["echo", "hello"]`.
+
+            schedule (`str`):
+                One of "@annually", "@yearly", "@monthly", "@weekly", "@daily", "@hourly", or a
+                CRON schedule expression (e.g., '0 9 * * 1' for 9 AM every Monday).
+
+            suspend (`bool`, *optional*):
+                If True, the scheduled Job is suspended (paused).
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py]: Does that mean one can create a cron job and pause it immediately?

File: src/huggingface_hub/hf_api.py [hunk 15]
```diff
@@ -10453,6 +10453,525 @@ def run_uv_job(
...
+
+            concurrency (`bool`, *optional*):
+                If True, multiple instances of this Job can run concurrently.
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py]:
```suggestion
                If True, multiple instances of this Job can run concurrently. Defaults to False.
```

File: src/huggingface_hub/hf_api.py [hunk 16]
```diff
@@ -10453,6 +10453,525 @@ def run_uv_job(
...
+            >>> command = ["python", "-c", "import torch; print(f"This code ran with the following GPU: {torch.cuda.get_device_name()}")"]
+            >>> schedule_job(image, command, flavor="a10g-small", schedule="@hourly")
+            ```
+
+        """
+        if flavor is None:
+            flavor = SpaceHardware.CPU_BASIC
+
+        # prepare job spec to send to HF Jobs API
+        job_spec: Dict[str, Any] = {
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py]: it would be good to have a centralized `_create_job_spec`. I feel that it's a 1:1 duplicate of `run_job` implementation.

File: src/huggingface_hub/hf_api.py [hunk 17]
```diff
@@ -10453,6 +10453,525 @@ def run_uv_job(
...
+                A valid user access token. If not provided, the locally saved token will be used, which is the
+                recommended authentication method. Set to `False` to disable authentication.
+                Refer to: https://huggingface.co/docs/huggingface_hub/quick-start#authentication.
+        """
+        if namespace is None:
+            namespace = self.whoami(token=token)["name"]
+        get_session().delete(
+            f"{self.endpoint}/api/scheduled-jobs/{namespace}/{scheduled_job_id}",
+            headers=self._build_hf_headers(token=token),
+        ).raise_for_status()
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py]:
```suggestion
        response = get_session().delete(
            f"{self.endpoint}/api/scheduled-jobs/{namespace}/{scheduled_job_id}",
            headers=self._build_hf_headers(token=token),
        )
        hf_raise_for_status(response)
```

File: src/huggingface_hub/hf_api.py (hunk 16 above)
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py]: (same comment for the create uv schedule one)

**Review by Wauplin — COMMENTED** [review]
Thanks for working on this @lhoestq!

I find it a bit sad to have 2 different routes to create/update Jobs and Schedule Jobs as it looks to me to be pretty much the same API with additional parameters. Let's keep it as it is since we are mimicking server API but at least let's factorize some logic to generate the payloads as there is quite some custom logic (especially uv scripts and `timeoutSeconds`)

**Comment by lhoestq** [comment]
> Quick UX question: what's the advantage of adding a dedicated scheduled subcommand vs allowing scheduling directly on hf jobs run with a --schedule flag, e.g:

> I find it a bit sad to have 2 different routes to create/update Jobs and Schedule Jobs as it looks to me to be pretty much the same API with additional parameters.

Well `hf jobs run` has `--detach` which doesn't make sense for scheduled jobs. But still I'm not against `hf jobs run --schedule ...` since it feels natural anyways.

I originally had `scheduled` subcommand for the delete/suspend/resume commands which only exist for scheduled jobs, and also because inspect/ps don't return the same format. 

Here is another suggestion using `--schedule` and without the subcommand. Note that delete/suspend/resume are only for scheduled jobs, and ps/inspect also includes scheduled jobs:

```
hf jobs run --schedule @hourly ubuntu echo hello world
hf jobs run --schedule "0 * * * *" ubuntu echo hello world
hf jobs ps
hf jobs inspect <id>
hf jobs delete <id>
hf jobs suspend <id>
hf jobs resume <id>
hf jobs uv run --schedule @weekly train.py
```

notes:
- for `inspect` I can auto-infer if it's a job or a scheduled job
- for `ps` I was thinking of merging the lists of jobs and scheduled jobs (instead of having two distinct lists) - and use the STATUS column for the status of the last job in case of a schedule. Let me know what you think !

```
ID                       IMAGE  COMMAND  CREATED             STATUS    SCHEDULE
------------------------ ------ -------- ------------------- --------- -----------
68b577d0c3308bfc7c640e72 ubuntu echo hey 2025-09-01 10:38:19 READY     @hourly
68b5777f37a21629b868b005 ubuntu echo hey 2025-09-01 10:38:24 RUNNING   * * * * *
68a6ee24716242753cc8591e ubuntu echo hey 2025-09-01 10:38:37 COMPLETED */5 * * * *
68b57787c3308bfc7c640e71 ubuntu echo hey 2025-09-01 10:45:59 RUNNING
```

File: src/huggingface_hub/hf_api.py (hunk 14 above)
Comment (lhoestq) [reply src/huggingface_hub/hf_api.py]: yep exactly

**Comment by julien-c** [comment]
no strong opinion on the `--schedule` flag thing. I'm not 100% sure this will be hugely used anyways so maybe we ship like this?

**Comment by hanouticelina** [comment]
i'm okay to ship like this! let's keep `hf jobs scheduled`

**Review by julien-c — COMMENTED** [review]
the UI for this has been shipped on the Hub (and it's quite neat): 
image

**Comment by lhoestq** [comment]
Alright let's ship with `hf jobs scheduled ...`. I took all your comments into account.

In particular I renamed `schedule_job()` -> `create_scheduled_job()` (same for uv) and I factorized the job_spec creation code and the UV command creation code.

**Review by Wauplin — APPROVED** [review]
Looks good to me! Thanks a lot for the changes and all the added docstrings/documentation. Nothing more to add :hugs:
